### PR TITLE
FastLoader have a white list mode

### PR DESCRIPTION
**Pull request by rogermichoud**
A new API has been provided for the FastLoader to
selectively choose the streams it wants to load. It also
adds checkpoint streams for each stream added to the
whitelist.

Along, some refactoring of fastLoader testing.

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 1]
```diff
@@ -90,6 +90,28 @@
...
+     * Enable whiteList mode where we only reconstruct
+     * the streams provided through this api.
+     * @param streamsToLoad
+     */
+    public FastObjectLoader addStreamsToLoad(List<UUID> streamsToLoad) {
+        whiteList = true;
+        this.streamsToLoad.addAll(streamsToLoad);
+
+        // Generate the streamsCheckpointId (we need to allow them as well)
+        streamsToLoad.forEach((streamId) ->
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: MINOR Remove the parentheses around the "streamId" parameter [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1611)

File: test/src/test/java/org/corfudb/recovery/Helpers.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    static VersionLockedObject getVersionLockedObject(CorfuRuntime cr, String streamName, Class type) {
+        CorfuCompileProxy cp = getCorfuCompileProxy(cr, streamName, type);
+        return cp.getUnderlyingObject();
+    }
+
+    static void assertThatMapIsNotBuilt(CorfuRuntime rt, String streamName, Class type) {
+        ObjectsView.ObjectID mapId = new ObjectsView.
+                ObjectID(CorfuRuntime.getStreamID(streamName), type);
+
+        assertThat(!rt.getObjectsView().getObjectCache().containsKey(mapId));
```
Comment (corfu-repo) [inline test/src/test/java/org/corfudb/recovery/Helpers.java]: BLOCKER Complete the assertion. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2970)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 2 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by bitmingw** [comment]
@rogermichoud We are having inconsistent interface for `addStreamToIgnore` and `addStreamsToLoad`. The ignore one takes table name and does a translate from table name to stream ID internally. I expect the add one to have a similar behavior: takes List<String> and do the transform.

**Comment by rogermichoud** [comment]
@bitmingw Will do the change. Also, there is another issue I have to fix before you can test it.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **d079081**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #973 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/X8TfRZcHPlQ47SPZ3JYsZQO0Kjl6hsWt)

**Comment by bitmingw** [comment]
@rogermichoud I think it basically works but ERROR logs are still printed out, which is annoying.

The problem can be described below.

```
loadMaps()
  - recoverRuntime()
    - processLogData()
      - updateCorfuObject()  <--- here
```

In `updateCorfuObject()` the first thing is constructing `logEntry` from `logData`. The deserailization will fail for a unknown type. Then the log `Cannot deserialize log entry` and its stack trace is printed out.

I purpose to suppress this ERROR message since it will confuse people.

**Comment by rogermichoud** [comment]
I will fix it this way. There are 2 cases:

1. No stream in the logData (it can contain multiple streams), will just completely ignore the logData.
2. We have a mix of streams (some are in the whitelist, some are not), I will go ahead with the deserialization and log an error if we cannot deserialize it. This means that a client wrote a transaction with a mix of tables we are able to deserialize and some that we are not.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **ef8dad1**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #973 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/4yeIAA3pq7rDmcJMne09ESH2ahMx9Nju)

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 3]
```diff
@@ -90,6 +90,26 @@
...
 
+    private boolean whiteList = false;
+    private List<UUID> streamsToLoad = new ArrayList<>();
+
+    /**
+     * Enable whiteList mode where we only reconstruct
+     * the streams provided through this api.
+     * @param streamsToLoad
+     */
+    public FastObjectLoader addStreamsToLoad(List<String> streamsToLoad) {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: What's the definition of white lists here, only load streams that are in the white list? 
Make sure that they are mutually exclusive (i.e. can't use both lists at the same time).

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 4]
```diff
@@ -232,20 +254,53 @@ public void updateStreamTails(long address, ILogData logData) {
      * @param entry entry to potentially apply
      * @return if we need to apply the entry.
      */
-    private boolean shouldEntryBeApplied(UUID streamId, SMREntry entry) {
+    private boolean shouldEntryBeApplied(UUID streamId, SMREntry entry, boolean isCheckpointEntry) {
+        // If white list mode, ignore everything that is not in the list or checkpoints
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: The comment is incorrect?

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 5]
```diff
@@ -232,20 +254,53 @@ public void updateStreamTails(long address, ILogData logData) {
...
+        if (whiteList && !streamsToLoad.contains(streamId)) {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: This is kind of superfluous, only the second clause is needed. Also, why negate/negate?

If (streamsToLoad.contains(streamId)) return true;

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 6]
```diff
@@ -232,20 +254,53 @@ public void updateStreamTails(long address, ILogData logData) {
...
+            return false;
+        }
+
         // We ignore the transaction stream ID because it is a raw stream
         // We don't want to create a Map for it.
         if (streamId == runtime.getObjectsView().TRANSACTION_STREAM_ID || streamsToIgnore.contains(streamId)) {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: You can either add this stream to the black list, or not include it in the white list.

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 7]
```diff
@@ -232,20 +254,53 @@ public void updateStreamTails(long address, ILogData logData) {
...
             return false;
         }
 
-        if (!streamsMetaData.containsKey(streamId) || entry.getEntry().getGlobalAddress() >
+        if (isCheckpointEntry || !streamsMetaData.containsKey(streamId) || entry.getEntry().getGlobalAddress() >
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: Just use static boolean isCheckPointEntry(ILogData logData) ?

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 8]
```diff
@@ -373,6 +373,14 @@ public static UUID getStreamID(String string) {
         return UUID.nameUUIDFromBytes(string.getBytes());
     }
 
+    public static UUID getCheckpointStreamIdFromId(UUID streamId) {
+        return getStreamID(streamId.toString() + "_cp");
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: Probably want to make the string lateral a static definition in streamview

File: test/src/test/java/org/corfudb/recovery/FastObjectLoaderTest.java [hunk 9]
```diff
@@ -813,233 +472,134 @@ public void canReadWithEntriesInterleavingCPS() throws Exception {
...
-        assertThatMapIsBuilt(rt1, rt2, "Map2", map2);
-        assertThatMapIsBuilt(rt1, rt2, "Map3", map3);
-        assertThatObjectCacheIsTheSameSize(rt1, rt2);
-
+        assertThatMapsAreBuilt(rt2);
+        assertThatObjectCacheIsTheSameSize(getDefaultRuntime(), rt2);
     }
 
     @Test
     public void ignoreStreamForRuntimeButNotStreamTails() throws Exception {
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/recovery/FastObjectLoaderTest.java]: What happens when you white list and ignore a stream?

**Comment by Maithem** [comment]
Also, rebase.

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 10]
```diff
@@ -90,6 +90,26 @@
     @Getter
     private boolean logHasNoCheckPoint = false;
 
+    private boolean whiteList = false;
```
[the remark this answers is not in the document]
Comment (rogermichoud) [reply runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: I used it as the knob for the mutual exclusion between the whitelist mode and the blacklist mode. If the client of the API wants to do whitelist but because of its logic he doesn't have any table in the provided streamsToLoad, it should still be a whitelist. 

As is, I could not infer from the streasToLoad if we are in whitelist mode with no streams given or if we are in blacklist mode.

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java (hunk 3 above)
Comment (rogermichoud) [reply runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: Will add in the comments that they are mutually exclusive. The way to enforce it is by using the whiteList flag.

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java (hunk 4 above)
Comment (rogermichoud) [reply runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: is poorly written.  I'll rephrase it. What I wanted to emphasize was that in this list there are both the streams passed by the client and their derived checkpoint streams.

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java (hunk 5 above)
Comment (rogermichoud) [reply runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: The logic is a bit more complicated. The last condition should always be checked, whether we are in whitelist or blacklist. It checks that we are not processing an entry that is already contained in a checkpoint we already processed.

This is why if streamsToLoad.contains(streamId) return true is actually not enough. I had this at the beginning, and it was not passing the tests.

I will re-work this function to make it more self-explanatory. I reckon it is a bit hard to read.

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java (hunk 7 above)
Comment (rogermichoud) [reply runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: I am too deep for that. I don't have access to the logData directly (or I would need to get it backward from the logEntry). Here I am processing an entry inside the CheckpointEntry. Easiest way to know if we are processing a checkpoint or a normal entry is to pass it along the parameters.

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java (hunk 8 above)
Comment (rogermichoud) [reply runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: sound right.

File: test/src/test/java/org/corfudb/recovery/FastObjectLoaderTest.java (hunk 9 above)
Comment (rogermichoud) [reply test/src/test/java/org/corfudb/recovery/FastObjectLoaderTest.java]: The expectation is that you should not be doing this. Modes are mutually exclusive, so adding a stream to ignore in the whitelist mode should be a prohibited operation. I added an exception when this happens.

**Comment by rogermichoud** [comment]
@Maithem I addressed the various comments. 

I agree it is debatable if the way I structured my testing is the best one, but I will let it like this for the 2 following reasons:
1. There is a lot of corner cases in these tests, refactoring them in bigger tests is kind of dangerous. I might miss some of these corner cases.
2. The advantage of having multiple small tests is that it is easier to understand what went wrong when some of them are not passing (and it happened many times). When refactoring the code, I saw that only the one related to checkpoints with extra data after them were failing. It was easier to locate where was the bug.

Beyond the style/small changes, I rework the shouldEntryBeApplied workflow to make it easier for the reader. Also added some test to ensure mutual exclusivity of whitelist/blacklist mode.

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 11]
```diff
@@ -90,6 +100,32 @@
...
+
+    /**
+     * Enable whiteList mode where we only reconstruct
+     * the streams provided through this api. In this mode,
+     * we will only process the streams present in streamsToLoad.
+     * All the others are ignored.
+     * @param streamsToLoad
+     */
+    public FastObjectLoader addStreamsToLoad(List<String> streamsToLoad) {
+        if (streamsToIgnore.size() != 0) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: MINOR Use isEmpty() to check whether the collection is empty or not. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1155)

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 12]
```diff
@@ -221,29 +265,85 @@ public void updateStreamTails(long address, ILogData logData) {
...
+     * the checkpoint. We must avoid to process them twice.
+     *
+     * This comes especially relevant when the operation order affects the end result.
+     * (e.g. clear() operation)
+     *
+     * @param streamId stream id to validate
+     * @param entry entry under scrutinization.
+     * @return if the entry is already part of the checkpoint we started from.
+     */
+    private boolean entryAlreadyContainedInCheckpoint(UUID streamId, SMREntry entry) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: MINOR Remove this use of "SMREntry"; it is deprecated. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **b49359b**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #973 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/ojOnvJP7YQDRSstZAYAj6Sxh0f6AIb5Y)

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: whiteList is set to true when you add a stream to a white list, so it can be inferred from the list size. Setting it to true even if the white list is empty is another issue.

Also, my initial point was its kind of redundant and it doesn't seem to work. Consider this example:

@Test
    public void whiteListMode() throws Exception {
        populateMaps(MORE, getDefaultRuntime(), CorfuTable.class, true, SOME);

        List<String> streamsToLoad = new ArrayList<>();
        streamsToLoad.add("Map1");
        streamsToLoad.add("Map3");

        CorfuRuntime rt2 = new CorfuRuntime(getDefaultConfigurationString()).connect();
        FastObjectLoader loader = new FastObjectLoader(rt2)
                .setDefaultObjectsType(CorfuTable.class);
        loader.addStreamToIgnore("a");
        loader.addStreamsToLoad(streamsToLoad);
        loader.loadMaps();

        assertThatMapsAreBuiltWhiteList(rt2, streamsToLoad);
    }

**Comment by rogermichoud** [comment]
@Maithem 
The two modes are mutually exclusive. The operation you mentioned is prohibited. These wrong usage of the loader (mixing modes) are tested in blackListAndWhiteListAreMutuallyExclusive.

Better be safe than sorry, I copied verbatim your example and it has the expected behavior, throwing a IllegalStateException: Cannot add a whitelist when there are already streams to ignore

Your concern on robustness and readibility is valid. The whiteMode flag actually reduce code in the addStreasToLoad method. Since an empty list is already initialized, I don't need to add extra logic to check if the list has been initialized (or use lombok lazy initialization that is sometimes a bit opaque).  I will argue that this is more code style than code functionality. Also, the alternative is to have a method isWhiteListModeEnable that infers from the list (and from some other mechanism to see if we passed a empty list) if we are in whiteListMode. I am not sure this would make anything more readable.

Anyway, as long as functionaly works and readibility is okay, I agree we should proceed.

**Comment by Maithem** [comment]
Ok, I see the test failing now. Didn't see the force-push.
